python/GafferSceneTest/SceneAlgoTest.py

```diff
@@ -1803,6 +1804,123 @@ def assertOptionHistory( optionName, nameSwitchInput, value ) :
+		script["expression"].setExpression( inspect.cleandoc(
+		"""
+		globals = parent["dot"]["in"]["globals"]
+		globals["option:render:camera"] = IECore.StringData( "expression" )
+		parent["dot"]["out"]["globals"] = globals
+		"""
+		) )
+
+		def runTest() :
+
+			history = GafferScene.SceneAlgo.history( script["dot"]["out"]["globals"] )
+
+			for plug, value in [
+				( script["dot"]["out"], "expression" ),
+				( script["dot"]["in"], "test" ),
+				( script["options"]["out"], "test" ),
+			] :
+				self.assertEqual( history.scene, plug )
+				self.assertLessEqual( len( history.predecessors ), 1 )
+				history = history.predecessors[0] if history.predecessors else None
+
+			self.assertIsNone( history )
+
+		# Test running the test while everything is already cached still works, and doesn't add any
+		# new entries to the cache
+		Gaffer.ValuePlug.clearHashCache()
+		runTest()
+		before = Gaffer.ValuePlug.hashCacheTotalUsage()
+		runTest()
+		self.assertEqual( Gaffer.ValuePlug.hashCacheTotalUsage(), before )
+
+		# Test that even the processes that aren't reading the cache still write to the cache, by
+		# making sure that a subsequent globalsHash doesn't need to do anything
+		Gaffer.ValuePlug.clearHashCache()
+		runTest()
+		with Gaffer.PerformanceMonitor() as pm :
+			script["dot"]["out"].globalsHash()
+		self.assertEqual( pm.combinedStatistics().hashCount, 0 )
+
+		# Test branching history with an expression using two input globals
+		# to produce our output
+		script["options2"] = GafferScene.StandardOptions()
+		script["options2"]["options"]["renderCamera"]["enabled"].setValue( True )
+		script["options2"]["options"]["renderCamera"]["value"].setValue( "other" )
+
+		script["expression"].setExpression( inspect.cleandoc(
+		"""
+		globals = parent["dot"]["in"]["globals"]
+		globals["option:render:camera"] = parent["options2"]["out"]["globals"].get( "option:render:camera" )
+		parent["dot"]["out"]["globals"] = globals
+		"""
+		) )
+
+		history = GafferScene.SceneAlgo.history( script["dot"]["out"]["globals"] )
+		self.assertEqual( len( history.predecessors ), 2 )
+		self.assertEqual( history.predecessors[0].scene, script["dot"]["in"] )
+		self.assertEqual( history.predecessors[1].scene, script["options2"]["out"] )
+
+		# Test that an expression using the value of a non-globals plug to modify the globals
+		# does not affect the history
+		script["cube"] = GafferScene.Cube()
+		script["cube"]["sets"].setValue( "foo" )
+
+		script["expression"].setExpression( inspect.cleandoc(
+		"""
+		globals = parent["dot"]["in"]["globals"]
+		globals["option:render:camera"] = IECore.StringData( "expression" )
+		globals["option:setNames"] = parent["cube"]["out"]["setNames"]
+		parent["dot"]["out"]["globals"] = globals
+		"""
+		) )
+
+		history = GafferScene.SceneAlgo.history( script["dot"]["out"]["globals"] )
+		self.assertEqual( len( history.predecessors ), 1 )
+		self.assertEqual( history.predecessors[0].scene, script["dot"]["in"] )
```

Comment:
Do we need to assert that `predecessors[0]` has no predecessors here? I think that's where we'd expect to find the unwanted `setNames` event?

Reply:
If the unwanted `setNames` event appeared, I'd expect to see it directly in `history.predecessors` like we see at L1904 of the previous test.

`predecessors[0]` would still have a legit predecessor as the globals arrive at `script["dot"]["in"]` from the upstream StandardOptions node.